### PR TITLE
Attach APIs to Signup Form

### LINKED ISSUE #54: Attach APIs to the Signup form
Right now submitting signup form doesn't do anything. 

It should signup the user and log him in!

### PR DESCRIPTION
**Pull request by niinpatel**
resolves #54

### CODE REVIEW

**Comment by niinpatel** [comment]
To-do list on this PR:

1. Add tests for LogoutButton.
2. Add tests for SignupAndLoginButtons.
3. Update tests for auth buttons checking if LogoutButton and SignupAndLoginButtons are rendered correctly. 
4. Update Tests for useSignupForm to check if Server API is being called. 
5. Mock Server APIs and check if useSignupForm handles errors and success.

**Comment by niinpatel** [comment]
Live Demo: 

https://www.useloom.com/share/f434fde59e0a48fa8a20819f224292e0

**Comment by niinpatel** [comment]
Alright, this PR is finally ready. It is a big pull-request. So, here's the summary of the changes I made. 

1. The `user` state in <App/> component is passed to the rest of apps using [ContextAPI](https://reactjs.org/docs/context.html) instead of prop drilling. I'm also passing functions called `login` and `logout` which basically set user state and remove user from state (respectively). 


2. As the title of this PR says, submitting signup form now actually calls the server API. This is done inside the useSignupForm hook.

3. If the signup was successful, it calls handleFormSuccess which sets jwt token returned by server in `localStorage` and login the user. 

4. If signup failed, it calls handleFormSuccess which displays the errors to user by in the form fields. 

5. The <App /> component, when rendered, checks if there is jwt token in localStorage. If valid, jwt exists, it will log in the user using jwt. If jwt was invalid or expired, it will remove it from localStorage.

6. Finally, I've changed the <AuthButtons /> so that if a user is logged in, it will show logout button and username. 
If no user is logged in, it will show signup and login buttons and their respective modals. 

Here's a preview of all of this: 



class component demo 1

File: client/src/App.js [hunk 1]
```diff
@@ -1,24 +1,62 @@
...
+    const token = localStorage.getItem('token');
+    if (!this.state.user && token) {
+      this.loginUsingJWT(token);
+    }
   }
 
+  loginUsingJWT = (token) => {
+    const decoded = decode(token);
+    const tokenIsValid = decoded !== null;
+    if (tokenIsValid) {
```
Comment (nimish-gupta) [inline client/src/App.js]: Try to use early return pattern
```suggestion
    if (!tokenIsValid) {
       this.removeTokenFromLocalStorage();
       return;
    }
     const expired = decoded.exp <= new Date() / 1000;
     if (expired) {
         this.removeTokenFromLocalStorage();
         return; 
     }
     this.login({ id: decoded.id, username: decoded.username });
```

File: client/src/App.js [hunk 2]
```diff
@@ -1,24 +1,62 @@
...
+        this.removeTokenFromLocalStorage();
+      } else {
+        this.login({ id: decoded.id, username: decoded.username });
+      }
+    } else {
+      this.removeTokenFromLocalStorage();
+    }
+  };
+
+  removeTokenFromLocalStorage = () => {
```
Comment (nimish-gupta) [inline client/src/App.js]: Remove extra curly braces

File: client/src/App.js [hunk 3]
```diff
@@ -1,24 +1,62 @@
...
+  };
+
+  logout = () => {
+    this.removeTokenFromLocalStorage();
+    this.setState({
+      user: null,
+    });
+  };
+
+  login = (user) => {
```
Comment (nimish-gupta) [inline client/src/App.js]: Remove extra curly braces

File: client/src/App.test.js [hunk 4]
```diff
@@ -1,14 +1,55 @@
...
+    const mockToken = sign({ username: 'abc', id: '123' }, 'secret', { expiresIn: 60 * 60 });
+    localStorage.setItem('token', mockToken);
+    const wrapper = mount(<App />);
+    expect(wrapper.state().user).toEqual({ username: 'abc', id: '123' });
+  });
+
+  test('should remove token from localstorage if token is invalid', () => {
+    localStorage.setItem('token', 'invalidtoken');
+    const wrapper = mount(<App />);
+    expect(wrapper.state().user).toBe(null);
```
Comment (nimish-gupta) [inline client/src/App.test.js]: wrap this to checkUnAuthenticatedUser() function and use it all the test cases
```suggestion
    const checkUnAuthenticatedUser = ({ state }) => {
       expect(wrapper.state().user).toBe(null);
       expect(localStorage.removeItem).toHaveBeenCalledWith('token');
    }
    
```

File: client/src/components/__tests__/SignupModal.test.js [hunk 5]
```diff
@@ -69,18 +73,64 @@ describe('input fields in form', () => {
...
+    usernameInpt.simulate('change', { target: { value: mockUser.username } });
+    emailInpt.simulate('change', { target: { value: mockUser.email } });
+    passwordInpt.simulate('change', { target: { value: mockUser.password } });
+    repeatPasswordInpt.simulate('change', { target: { value: '123456' } });
+
+    wrapper.find('form').simulate('submit');
+
+    return Promise.resolve().then(() => {
+      expect(signupAPI).toHaveBeenCalledWith(mockUser);
+      expect(handleSuccess).toHaveBeenCalledWith(mockUser);
```
Comment (nimish-gupta) [inline client/src/components/__tests__/SignupModal.test.js]: Here done should be called if it is promise

Comment (niinpatel) [reply client/src/components/__tests__/SignupModal.test.js]: If we return a promise from test, then jest waits for the promise. No need to call done. 

> If your code uses promises, there is a simpler way to handle asynchronous tests. Just return a promise from your test, and Jest will wait for that promise to resolve. If the promise is rejected, the test will automatically fail.

https://jestjs.io/docs/en/asynchronous

**Comment by niinpatel** [comment]
@nimish-gupta , made the changes requested so far.

**Comment by niinpatel** [comment]
Merging this PR.

**Pull request merged**